src/block_sync/client.rs

```diff
@@ -44,43 +116,496 @@ impl<N: Network> BlockSyncClient<N> {
+
+                        if response.highest_qc.is_correct(block_tree)?
+                            && safe_qc(&response.highest_qc, block_tree, self.config.chain_id)?
+                        {
+                            update_block_tree(
+                                &response.highest_qc,
+                                block_tree,
+                                &self.event_publisher,
+                            )?;
+                        }
+                    }
+                }
+                Err(BlockSyncResponseReceiveError::Disconnected)
+                | Err(BlockSyncResponseReceiveError::Timeout) => {
+                    // Check if the server's commitment to providing blocks at least up to a given height was observed.
+                    // If not, blacklist the server.
+                    let min_blocks_expected = *self
+                        .block_sync_client_state
+                        .available_sync_servers
+                        .get(peer)
+                        .unwrap()
+                        - init_highest_committed_block_height;
+
+                    if blocks_synced < min_blocks_expected {
+                        self.block_sync_client_state
+                            .blacklist_sync_server(peer.clone(), self.config.blacklist_expiry_time)
+                    }
+
+                    Event::EndSync(EndSyncEvent {
+                        timestamp: SystemTime::now(),
+                        peer: *peer,
+                        blocks_synced,
+                    })
+                    .publish(&self.event_publisher);
+                    return Ok(());
+                }
+            }
+        }
+    }
 }
 
-/// Immutable parameters that define the behaviour of the [BlockSyncClient].
+/// Immutable parameters that define the behaviour of the [`BlockSyncClient`]:
+/// 1. Chain ID of the target blockchain,
+/// 2. Max. number of blocks per block sync request,
+/// 3. Timeout for waiting for a single block sync response,
+/// 4. Time after which a blacklisted sync server should be removed from the blacklist.
+/// 5. By how many views a QC received via [`AdvertiseQC`] must be ahead of the current
+///    view in order to trigger sync (event-based sync trigger).
+/// 6. How much time needs to pass without any progress (i.e., updating the highestQC)
+///    or sync attempts in order to trigger sync (timeout-based sync trigger).
 pub(crate) struct BlockSyncClientConfiguration {
     pub(crate) chain_id: ChainID,
     pub(crate) request_limit: u32,
     pub(crate) response_timeout: Duration,
+    pub(crate) blacklist_expiry_time: Duration,
+    pub(crate) block_sync_trigger_min_view_difference: u64,
+    pub(crate) block_sync_trigger_timeout: Duration,
 }
 
 struct BlockSyncClientState {
-    sync_servers: HashMap<VerifyingKey, u64>,
-    blacklist: HashSet<VerifyingKey>,
-    // TODO: fields for timeout-based sync trigger
+    available_sync_servers: HashMap<VerifyingKey, BlockHeight>,
```

Comment:
Replicas are never removed from this HashMap unless they are blacklisted, so this HashMap can grow forever. We should preferably clear this HashMap of replicas NOT in the set "committed validators + potential validators" whenever appropriate, perhaps every time we update `highest_qc_view`.

We should also consider caching the set "committed validators + potential validators" instead of computing it every time we receive an `AdvertiseBlock`.

Reply:
Extracted into standalone issue #43. Will be resolved in a later PR. This PR should focus on the HotStuff subprotocol.